File: functions/Visualization_functions/F_Visualize_Ex3.py
import numpy as np
import matplotlib.pyplot as plt

def F_Visualize_Ex3(Y_ens, S_obs, S_true, S_ens, xPlus_3d, Sll_2d, Cxx_3d, Cxpxp_3d):
    clr_true = 'k'
    marker_true = '.'
    line_true = 'k-'
    msize_true = 15

    clr_obs = 'k'
    marker_obs = '^'
    line_obs = 'k-'
    msize_obs = 4

    clr_mod = '#d10000'
    marker_mod = 'o'
    line_mod = '--'
    clr_mod2 = '#d10000'

    clr_upd = '#527dea'
    marker_upd = 'o'
    line_upd = ':'
    clr_upd2 = 'w'

    clr_ens = (0.8, 0.8, 0.8)
    line_ens = '-'

    lw = 2
    msize = 6
    fsize = 12
    opt_save = True
    opt_saveData = False

    Ne = Y_ens.shape[1]
    sim = [1, 24]  # simulation phase of model

    # ---- Visualizations ----

    # truth and observations
    plt.figure()
    plt.gca().tick_params(labelsize=fsize)
    for nn in range(Ne):
        plt.plot(Y_ens[:, nn], line_ens, color=clr_ens, linewidth=lw)
    h2, = plt.plot(S_obs, line_obs, marker=marker_obs, markeredgecolor=clr_obs,
                   markerfacecolor=clr_obs, linewidth=lw, markersize=msize)
    h3, = plt.plot(S_true, line_true, marker=marker_true, markeredgecolor=clr_true,
                   markerfacecolor=clr_true, linewidth=lw, markersize=msize_true)
    plt.legend([h3, h2], ['Truth', 'Observation'])
    plt.xlabel('time')
    plt.xlim([-0.5, (sim[1] - sim[0] + 1) + 0.5])
    plt.ylabel('S')
    plt.title('Ground truth and observation ensemble')
    plt.grid(True)
    if opt_save:
        plt.savefig('figures/Ex3_Obs.jpg', dpi=150)
    plt.show()

    # model simulation (open loop)
    plt.figure()
    plt.gca().tick_params(labelsize=fsize)
    for nn in range(Ne):
        plt.plot(S_ens[:, nn], line_ens, color=clr_ens, linewidth=lw)
    h2, = plt.plot(np.mean(S_ens, axis=1), line_mod, marker=marker_mod, markeredgecolor=clr_mod,
                   markerfacecolor=clr_mod2, linewidth=lw, color=clr_mod, markersize=msize)
    plt.legend([h2], ['Model mean'])
    plt.xlabel('time')
    plt.xlim([-0.5, (sim[1] - sim[0] + 1) + 0.5])
    plt.ylabel('S')
    plt.title('Water storage estimates - OL')
    plt.grid(True)
    if opt_save:
        plt.savefig('figures/Ex3_OL.jpg', dpi=150)
    plt.show()

    # DA results
    length_sim = sim[1] - sim[0] + 1
    xPlus_ensMean = np.mean(xPlus_3d[0, :, :], axis=0)

    plt.figure()
    plt.gca().tick_params(labelsize=fsize)
    for nn in range(Ne):
        plt.plot(xPlus_3d[0, nn, :], line_ens, color=clr_ens, linewidth=lw)
    h2, = plt.plot(xPlus_ensMean, line_upd, marker=marker_upd, markeredgecolor=clr_upd,
                   markerfacecolor=clr_upd, linewidth=lw, color=clr_upd, markersize=msize)
    plt.legend([h2], ['Model mean'])
    plt.xlabel('time')
    plt.xlim([-0.5, length_sim + 0.5])
    plt.ylabel('S')
    plt.title('Water storage estimates - DA')
    plt.grid(True)
    if opt_save:
        plt.savefig('figures/Ex3_DA.jpg', dpi=150)
    plt.show()

    # ensemble prediction
    plt.figure()
    plt.gca().tick_params(labelsize=fsize)
    plt.plot(np.mean(S_ens, axis=1), line_mod, marker=marker_mod, markeredgecolor=clr_mod,
             markerfacecolor=clr_mod2, linewidth=lw, color=clr_mod, markersize=msize)
    plt.plot(xPlus_ensMean, line_upd, marker=marker_upd, markeredgecolor=clr_upd,
             markerfacecolor=clr_upd2, linewidth=lw, color=clr_upd, markersize=msize)
    plt.plot(S_obs, line_obs, marker=marker_obs, markeredgecolor=clr_obs,
             markerfacecolor=clr_obs, linewidth=lw, markersize=msize)
    plt.plot(S_true, line_true, marker=marker_true, markeredgecolor=clr_true,
             markerfacecolor=clr_true, linewidth=lw, markersize=msize_true)
    plt.legend(['OL', 'DA', 'Observation', 'Truth'])
    plt.xlabel('time')
    plt.xlim([-0.5, length_sim + 0.5])
    plt.ylabel('S')
    plt.title('Water storage estimates (ensemble average)')
    plt.grid(True)
    if opt_save:
        plt.savefig('figures/Ex3_EnsAv.jpg', dpi=150)
    plt.show()

    # empirical standard deviations
    plt.figure()
    plt.gca().tick_params(labelsize=fsize)
    plt.plot(np.sqrt(Sll_2d), line_obs, marker=marker_obs, markeredgecolor=clr_obs,
             markerfacecolor=clr_obs, linewidth=lw, markersize=msize)

    test1 = np.zeros(length_sim)
    test1[:] = Cxx_3d[0, 0, :]
    plt.plot(np.sqrt(test1), line_mod, marker=marker_mod, markeredgecolor=clr_mod,
             markerfacecolor=clr_mod2, linewidth=lw, color=clr_mod, markersize=msize)

    test2 = np.zeros(length_sim)
    test2[:] = Cxpxp_3d[0, 0, :]
    plt.plot(np.sqrt(test2), line_upd, marker=marker_upd, markeredgecolor=clr_upd,
             markerfacecolor=clr_upd2, linewidth=lw, color=clr_upd, markersize=msize)

    plt.xlabel('time')
    plt.xlim([-0.5, length_sim + 0.5])
    plt.ylabel('S sigma')
    plt.title('Uncertainty (based on ensemble spread)')
    plt.grid(True)
    plt.legend(['Observations', 'Model Prediction', 'Model Update'])
    if opt_save:
        plt.savefig('figures/Ex3_Uncert.jpg', dpi=150)
    plt.show()

File: functions/Visualization_functions/F_Visualize_Ex3.m
function F_Visualize_Ex3(Y_ens,S_obs,S_true,S_ens,xPlus_3d,Sll_2d,Cxx_3d,Cxpxp_3d)

clr_true = 'k'; marker_true = '.'; line_true = 'k-';  msize_true = 15;
clr_obs  = 'k'; marker_obs  = '^'; line_obs  = 'k-';  msize_obs = 4;
clr_mod  = '#d10000'; marker_mod  = 'o'; line_mod  = '--';  clr_mod2 = '#d10000';
clr_upd  = '#527dea'; marker_upd  = 'o'; line_upd  = ':'; clr_upd2 = 'w';
clr_ens  = [0.8 0.8 0.8];          line_ens  = '-';
lw = 2;
msize = 6;
fsize = 12;
opt_save = 1;    
opt_saveData = 0;
Ne = size(Y_ens,2);
sim = [1; 24]; % simulation phase of model


%% Visualizations

% truth and observations
figure; hold on;
set(gca,'FontSize',fsize);
for nn=1:Ne
    h1 = plot(Y_ens(:,nn),line_ens,'Color',clr_ens,'LineWidth',lw);
end
h2 = plot(S_obs,line_obs,'Marker',marker_obs,'MarkerEdgeColor',clr_obs,'MarkerFaceColor',clr_obs,'LineWidth',lw,'MarkerSize',msize);
h3 = plot(S_true,line_true,'Marker',marker_true,'MarkerEdgeColor',clr_true,'MarkerFaceColor',clr_true,'LineWidth',lw,'MarkerSize',msize_true);
legend([h3 h2 h1], 'Truth', 'Observation', 'Ensemble (Obs)')
xlabel('time'); xlim([0-0.5 (sim(2)-sim(1)+1)+0.5]);
ylabel('S')
title('Ground truth and observation ensemble');
box on
if(opt_save==1) print(gcf,'figures/Ex3_Obs','-djpeg'); end

% model simulation (open loop)
figure; hold on;
set(gca,'FontSize',fsize)
for nn=1:Ne
    h1 = plot(S_ens(:,nn),line_ens,'Color',clr_ens,'LineWidth',lw);
end
h2 = plot(mean(S_ens,2),line_mod,'Marker',marker_mod,'MarkerEdgeColor',clr_mod,'MarkerFaceColor',clr_mod2,'LineWidth',lw,'Color',clr_mod,'MarkerSize',msize);
legend([h2 h1], 'Model mean', 'Ensemble (Model)')
xlabel('time'); xlim([0-0.5 (sim(2)-sim(1)+1)+0.5]);
ylabel('S')
title('Water storage estimates - OL');
box on
if(opt_save==1) print(gcf,'figures/Ex3_OL','-djpeg'); end

% DA results
id = logical(ones(1,sim(2)-sim(1)+1));
xPlus_ensMean = zeros(1,sim(2)-sim(1)+1); xPlus_ensMean(id) = mean(xPlus_3d,2);
figure; hold on;
set(gca,'FontSize',fsize)
for nn=1:Ne
    h1 = plot(squeeze(xPlus_3d(1,nn,:)),line_ens,'Color',clr_ens,'LineWidth',lw);
end
h2 = plot(xPlus_ensMean(1,:),line_upd,'Marker',marker_upd,'MarkerEdgeColor',clr_upd,'MarkerFaceColor',clr_upd,'LineWidth',lw,'Color',clr_upd,'MarkerSize',msize);
legend([h2 h1], 'Model mean', 'Ensemble (Model)')
xlabel('time'); xlim([0-0.5 (sim(2)-sim(1)+1)+0.5]);
ylabel('S')
title('Water storage estimates - DA');
box on
if(opt_save==1) print(gcf,'figures/Ex3_DA','-djpeg'); end

% ensemble prediction
figure; hold on;
set(gca,'FontSize',fsize)
plot(mean(S_ens,2),line_mod,'Marker',marker_mod,'MarkerEdgeColor',clr_mod,'MarkerFaceColor',clr_mod2,'LineWidth',lw,'Color',clr_mod,'MarkerSize',msize);
plot(xPlus_ensMean(1,:),line_upd,'Marker',marker_upd,'MarkerEdgeColor',clr_upd,'MarkerFaceColor',clr_upd2,'LineWidth',lw,'Color',clr_upd,'MarkerSize',msize);
plot(S_obs,line_obs,'Marker',marker_obs,'MarkerEdgeColor',clr_obs,'MarkerFaceColor',clr_obs,'LineWidth',lw,'MarkerSize',msize);
plot(S_true,line_true,'Marker',marker_true,'MarkerEdgeColor',clr_true,'MarkerFaceColor',clr_true,'LineWidth',lw,'MarkerSize',msize_true);
legend('OL','DA','Observation','Truth')
xlabel('time'); xlim([0-0.5 (sim(2)-sim(1)+1)+0.5]);
ylabel('S')
title('Water storage estimates (ensemble average)');
box on
if(opt_save==1) print(gcf,'figures/Ex3_EnsAv','-djpeg'); end

% empirical standard deviations
% observations - model prediction - model update
figure; hold on;
set(gca,'FontSize',fsize)
plot(sqrt(Sll_2d),line_obs,'Marker',marker_obs,'MarkerEdgeColor',clr_obs,'MarkerFaceColor',clr_obs,'LineWidth',lw,'MarkerSize',msize);
IN = logical(ones(24,1)); test1 = zeros(24,1);
test1(IN)=Cxx_3d(1,1,:);
plot(sqrt(test1),line_mod,'Marker',marker_mod,'MarkerEdgeColor',clr_mod,'MarkerFaceColor',clr_mod2,'LineWidth',lw,'Color',clr_mod,'MarkerSize',msize);
test2 = zeros(24,1); test2(IN)=Cxpxp_3d(1,1,:);
plot(sqrt(test2),line_upd,'Marker',marker_upd,'MarkerEdgeColor',clr_upd,'MarkerFaceColor',clr_upd2,'LineWidth',lw,'Color',clr_upd,'MarkerSize',msize);
xlabel('time'); xlim([0-0.5 (sim(2)-sim(1)+1)+0.5]);
ylabel('S sigma')
title('Uncertainty (based on ensemble spread)');
box on
legend('Observations','Model Prediction','Model Update')
if(opt_save==1) print(gcf,'figures/Ex3_Uncert','-djpeg'); end

end

